src/core/storage/storage.js

```diff
@@ -32,65 +88,162 @@ export default class GlobalStorage {
    */
-  get (key) {}
+  get (key) {
+    if (!this.moduleDataContainer.get(key)) {
+      return this.moduleDataContainer.get(key);
+    }
+    return this.moduleDataContainer.get(key).raw();
+  }
 
   /**
    * Get the current state for all key/value pairs in storage
    *
-   * @return {Object} mapping from key to value representing the
-   *                  current state
+   * @return {Map<string, ModuleData>} mapping from key to value representing the current state
    */
-  getAll () {}
+  getAll () {
+    const entries = new Map();
+    this.moduleDataContainer.forEach((value, key) => {
+      entries.set(key, this.get(key));
+    });
+    return entries;
+  }
 
   /**
-   * Remove the data in storage with the given key to the
-   * provided data
+   * Remove the data in storage with the given key
    *
    * @param {string} key The storage key to delete
    */
-  delete (key) {}
+  delete (key) {
+    this.moduleDataContainer.delete(key);
+    this.persistentStorageBuffer.delete(key);
+  }
 
   /**
    * Returns the query parameters to encode the current state
    *
    * @return {string} The query parameters for a page link with the
    *                  current state encoded
-   *                  e.g. ?query=all&context=%7Bkey:'hello'%7D
+   *                  e.g. query=all&context=%7Bkey:'hello'%7D
    */
-  getUrlWithCurrentState () {}
+  getUrlWithCurrentState () {
+    return this.persistentStorageBuffer.getUrlWithCurrentState();
+  }
 
   /**
    * Adds a listener to the given module for a given event
    *
    * @param {string} event The event to listen to, e.g. ‘update’
-   * @param {string} module The module key to listen for e.g. Pagination
+   * @param {string} moduleId The module key to listen for e.g. Pagination
    * @param {Function} callback The callback when an event is called,
    *                   function is given the data if relevant
    */
-  on (event, module, callback) {}
+  on (event, moduleId, callback) {
+    const moduleData = this.moduleDataContainer.get(moduleId);
+    if (moduleData === undefined) {
+      if (this.futureListeners.get(moduleId) === undefined) {
+        this.futureListeners.set(moduleId, []);
+      }
+      this.futureListeners.get(moduleId).push({ event, callback });
+      return;
+    }
+
+    moduleData.on(event, callback);
+  }
+
+  /**
+   * Removes a listener for the given moduleId for a given event
+   *
+   * @param {string} event The event being listened to, e.g. ‘update’
+   * @param {string} moduleId The module key being listened for e.g. Pagination
+   * @param {Function} callback The callback to be removed
+   */
+  off (event, moduleId, callback) {
+    const moduleData = this.moduleDataContainer.get(moduleId);
+    if (moduleData === undefined) {
+      if (this.futureListeners.get(moduleId) !== undefined) {
+        this.futureListeners.get(moduleId).pop(); // Is this right?
+      }
+      return;
+    }
+
+    moduleData.off(event, callback);
```

Comment:
I think this removes all callbacks for the moduleData, since this ends up calling event emitter's off(), which doesn't even take a callback arg? (ran into this a while back)

Reply:
Module data is out, using a different removeListener system!